File: sample-conference-videochat/sample-conference-videochat/Components/Calls/ConferenceViewController/ConferenceViewController.m
//
//  ConferenceViewController.m
//  sample-conference-videochat
//
//  Created by Injoit on 24.03.2021.
//  Copyright © 2021 Quickblox. All rights reserved.
//

#import "ConferenceViewController.h"
#import "OpponentsMediaViewController.h"

@interface ConferenceViewController ()
@property (strong, nonatomic) UIBarButtonItem *chatBarButtonItem;
@property (nonatomic, strong) id observerWillResignActive;
@property (nonatomic, strong) id observerWillActive;
@end

@implementation ConferenceViewController

- (void)viewWillDisappear:(BOOL)animated {
    [super viewWillDisappear:animated];
    
    [self setupNavigationBarWillAppear:NO];
    [self invalidateHideToolbarTimer];
    
    NSNotificationCenter *defaultCenter = [NSNotificationCenter defaultCenter];
    if (self.observerWillResignActive) {
        [defaultCenter removeObserver:(self.observerWillResignActive)];
    }
    if (self.observerWillActive) {
        [defaultCenter removeObserver:(self.observerWillActive)];
    }
}

- (void)didSetMuteVideo:(BOOL)muteVideo {
    self.session.localMediaStream.videoTrack.enabled = !muteVideo;
    [self.participants participantWithId:self.participants.localId].isCameraEnabled = !muteVideo;
    self.swapCamera.userInteractionEnabled = self.session.localMediaStream.videoTrack.enabled;
}

- (void)setupSession {
    if (!QBSession.currentSession.currentUser) {
        return;
    }
    QBUUser *currentUser = QBSession.currentSession.currentUser;

    
    // creating session
    NSString *conferenceID = self.conferenceSettings.conferenceInfo.conferenceID;
    self.session = [[QBRTCConferenceClient instance] createSessionWithChatDialogID:conferenceID conferenceType:QBRTCConferenceTypeVideo];
    if (!self.session) {
        return;
    }
    [self.participants addParticipantWithId:@(currentUser.ID) fullName:currentUser.name];
    
    NSNotificationCenter *center = NSNotificationCenter.defaultCenter;
    __weak __typeof(self)weakSelf = self;
    self.observerWillActive = [center addObserverForName:UIApplicationDidBecomeActiveNotification
                                                  object:nil
                                                   queue:NSOperationQueue.mainQueue
                                              usingBlock:^(NSNotification * _Nonnull note) {
        
        if ([weakSelf.participants participantWithId:weakSelf.participants.localId].isCameraEnabled != YES) {return;}
        weakSelf.session.localMediaStream.videoTrack.enabled = YES;
    }];
    
    self.observerWillResignActive = [center addObserverForName:UIApplicationWillResignActiveNotification
                                                        object:nil
                                                         queue:NSOperationQueue.mainQueue
                                                    usingBlock:^(NSNotification * _Nonnull note) {
        weakSelf.session.localMediaStream.videoTrack.enabled = NO;
    }];
}

- (void)configureNavigationBarItems {
    self.title = [self.chatManager.storage dialogWithID:self.conferenceSettings.conferenceInfo.chatDialogID].name;
    
    self.navigationItem.leftBarButtonItem = [[UIBarButtonItem alloc] initWithImage:[UIImage imageNamed:@"menu_chat"]
                                                                             style:UIBarButtonItemStylePlain
                                                                            target:self
                                                                            action:@selector(didTapChat:)];
    self.navigationItem.leftBarButtonItem.tintColor = UIColor.whiteColor;
    self.navigationItem.leftBarButtonItem.enabled = NO;
    
    self.navigationItem.rightBarButtonItem = [[UIBarButtonItem alloc] initWithImage:[UIImage imageNamed:@"members_call"]
                                                                              style:UIBarButtonItemStylePlain
                                                                             target:self
                                                                             action:@selector(didTapMembers:)];
    self.navigationItem.rightBarButtonItem.tintColor = UIColor.whiteColor;
    self.navigationItem.rightBarButtonItem.enabled = NO;
}

- (void)session:(__kindof QBRTCBaseSession *)session startedConnectingToUser:(NSNumber *)userID {
    if (session != self.session) {
        return;
    }
    [self addToCollectionUserWithID:userID];
}

- (void)addNewPublisher:(QBUUser *)user {
    [self.participants addParticipantWithId:@(user.ID) fullName:user.fullName];
    [self reloadContent];
    if ([self.callViewControllerDelegate respondsToSelector:@selector(callVCdidAdd:NewPublisher:)]) {
        [self.callViewControllerDelegate callVCdidAdd:YES NewPublisher:@(user.ID)];
    }
}

- (void)removeUserFromCollection:(NSNumber *)userID {
    [self.participants removeParticipantWithId:userID];
    if ([self.callViewControllerDelegate respondsToSelector:@selector(callVCdidAdd:NewPublisher:)]) {
        [self.callViewControllerDelegate callVCdidAdd:NO NewPublisher:userID];
    }
}

- (void)setupAudioVideoEnabledCell:(ConferenceUserCell *)cell forUserID:(NSNumber *)userID {
    if (userID == self.participants.localId) {
        self.session.localMediaStream.videoTrack.enabled = !self.muteVideo;
    }
    cell.isMuted = [self.participants participantWithId:userID].isEnabledSound;
}

// MARK: Internal Methods
- (void)didTapMembers:(UIBarButtonItem *)sender {
    NSString *dialogID = self.conferenceSettings.conferenceInfo.chatDialogID;
    OpponentsMediaViewController *opponentsMediaViewController = [[OpponentsMediaViewController alloc] initWithDialogID:dialogID users:self.participants.participants];
    self.callViewControllerDelegate = opponentsMediaViewController;

    __weak __typeof(self)weakSelf = self;
    [opponentsMediaViewController setDidPressMuteUser:^(BOOL isMuted, NSNumber * _Nonnull userID) {
        __typeof(weakSelf)strongSelf = weakSelf;
        QBRTCAudioTrack *audioTrack = [strongSelf.session remoteAudioTrackWithUserID:userID];
        audioTrack.enabled = !isMuted;
        [strongSelf.participants participantWithId:userID].isEnabledSound = !isMuted;
        [strongSelf reloadContent];
    }];

    self.session.localMediaStream.videoTrack.enabled = NO;
    [self.navigationController pushViewController:opponentsMediaViewController animated:YES];
}

@end

File: sample-conference-videochat-swift/sample-conference-videochat-swift/Components/Calls/ConferenceViewController/ConferenceViewController.swift
//
//  ConferenceViewController.swift
//  sample-conference-videochat-swift
//
//  Created by Injoit on 16.08.2020.
//  Copyright © 2020 QuickBlox. All rights reserved.
//

import UIKit
import Quickblox
import QuickbloxWebRTC
import SVProgressHUD

struct UserTracksStates {
    var isCameraEnabled: Bool
    var isEnabledSound: Bool
}

class ConferenceViewController: BaseCallViewController {
    private var usersStates:[UInt: UserTracksStates] = [:]
    private var chatDialogName = ""
    private var currentConferenceUserID = Profile().ID
    override var conferenceSettings: ConferenceSettings {
        didSet {
            self.chatDialogName = chatManager.storage.dialog(withID: conferenceSettings.conferenceInfo.chatDialogID)!.name!
        }
    }
    
    private lazy var membersItem = UIBarButtonItem(image: UIImage(named: "members_call"),
                                                   style: .plain,
                                                   target: self,
                                                   action: #selector(didTapMembers(_:)))
    
    override var muteVideo: Bool {
        didSet {
            session?.localMediaStream.videoTrack.isEnabled = !muteVideo
            usersStates[currentConferenceUserID]?.isCameraEnabled = !muteVideo
            swapCamera.isUserInteractionEnabled = !muteVideo
        }
    }
    
    // MARK: Overrides
    override func viewWillDisappear(_ animated: Bool) {
        super.viewWillDisappear(animated)
        
        setupNavigationBarWillAppear(false)
        invalidateHideToolbarTimer()
    }
    
    //MARK - Setup
    override func setupDelegates() {
        QBRTCConferenceClient.instance().add(self)
        QBRTCAudioSession.instance().addDelegate(self)
    }
    
    override func setupSession() {
        // creating session
        session = QBRTCConferenceClient.instance().createSession(withChatDialogID: conferenceSettings.conferenceInfo.conferenceID,
                                                                 conferenceType: .video)
        
        guard session != nil, let currentConferenceUser = createConferenceUser(userID: Profile().ID) else {
            return
        }
        let currentUserTracksStates = UserTracksStates(isCameraEnabled: false, isEnabledSound: true)
        usersStates[currentConferenceUserID] = currentUserTracksStates
        users = [currentConferenceUser]
    }
    
    override func configureNavigationBarItems() {
        title = chatDialogName
        navigationItem.rightBarButtonItem = membersItem
        
        navigationItem.leftBarButtonItem = UIBarButtonItem(image: UIImage(named: "menu_chat"),
                                                           style: .plain,
                                                           target: self,
                                                           action: #selector(didTapChat(_:)))
        navigationItem.leftBarButtonItem?.tintColor = .white
        navigationItem.leftBarButtonItem?.isEnabled = false
        navigationItem.rightBarButtonItem?.tintColor = .white
        navigationItem.rightBarButtonItem?.isEnabled = false
    }
    
    override func session(_ session: QBRTCBaseSession, startedConnectingToUser userID: NSNumber) {
        guard session == self.session else {
            return
        }
        // adding user to the collection
        addToCollectionUser(withID: userID)
    }
    
    override func addNewPublisher(_ user: ConferenceUser) {
        users.insert(user, at: 0)
        usersStates[user.userID] = UserTracksStates(isCameraEnabled: false, isEnabledSound: true)
        reloadContent()
    }
    
    override func removeUserFromCollection(_ userID: NSNumber) {
        if let index = users.index(where: { $0.userID == userID.uintValue }) {
            users.remove(at: index)
        }
    }
    
    override func setupAudioVideoEnabledCell(_ cell: ConferenceUserCell, forUserID userID: UInt) {
        if userID == currentConferenceUserID {
            session?.localMediaStream.videoTrack.isEnabled = !muteVideo
        }
        
        if let isMute = usersStates[userID]?.isEnabledSound  {
            cell.unMute = isMute
        }
    }
    
    // MARK: Internal Methods
    @objc private func didTapMembers(_ sender: UIBarButtonItem) {
        guard let chatInfoViewController = ScreenFactory().makeInfoUsersOutput() else { return }
        chatInfoViewController.dialogID = conferenceSettings.conferenceInfo.chatDialogID
        self.callViewControllerDelegate = chatInfoViewController
        chatInfoViewController.action = ChatAction.infoFromCall
        let qbUsers = users.compactMap{ chatManager.storage.user(withID: $0.userID) }
        chatInfoViewController.users = qbUsers
        chatInfoViewController.usersStates = usersStates
        chatInfoViewController.didPressMuteUser = { [weak self] (isMuted, userId) in
            guard let self = self else {
                return
            }
            let userID = NSNumber(value: userId)
            let audioTrack = self.session?.remoteAudioTrack(withUserID: userID)
            audioTrack?.isEnabled = !isMuted
            self.usersStates[userId]?.isEnabledSound = !isMuted
            self.reloadContent()
        }
        session?.localMediaStream.videoTrack.isEnabled = false
        navigationController?.pushViewController(chatInfoViewController, animated: true)
    }
}
